### PR TITLE
Example

### PR DESCRIPTION
**Pull request by yalh76**
## Problem

- *#35 not working*

## Solution

- *Apply example_ynh*

## PR Status

- [x] Code finished and ready to be reviewed/tested
- [x] The fix/enhancement were manually tested (if applicable)

## Automatic tests

Automatic tests can be triggered on https://ci-apps-dev.yunohost.org/ *after creating the PR*, by commenting "!testme", "!gogogadgetoci" or "By the power of systemd, I invoke The Great App CI to test this Pull Request!". (N.B. : for this to work you need to be a member of the Yunohost-Apps organization)

### CODE REVIEW

**Comment by yalh76** [comment]
!testme

**Comment by autra** [comment]
@yalh76 I'm actually already working on fixing the node version. That being said, your PR is much more than that. Do you think it'll be possible to split it in several smaller changes? 😸

**Comment by yalh76** [comment]
> @yalh76 I'm actually already working on fixing the node version. That being said, your PR is much more than that. Do you think it'll be possible to split it in several smaller changes? 😸

Yes, it's actually applying last example_ynh to make it easier to maintain.

If you prefer to split it in several smaller changes, feel free to do

**Comment by yalh76** [comment]
!testme

**Comment by yalh76** [comment]
!testme

**Comment by yalh76** [comment]
!testme

**Comment by yalh76** [comment]
!testme

**Comment by yalh76** [comment]
!testme

**Comment by yalh76** [comment]
!testme

**Comment by autra** [comment]
(just a quick message to tell you I haven't forgotten this ;-) I'll review/merge as soon as I can !)

**Comment by yalh76** [comment]
!testme

**Comment by yalh76** [comment]
!testme

**Comment by yalh76** [comment]
!testme

File: conf/database.yml [hunk 1]
```diff
@@ -2,8 +2,8 @@ postgresql: &postgresql
   adapter: postgresql
   host: "localhost"
   port: 5432
-  username: "{{ app }}"
-  password: "{{ db_pass }}"
+  username: "__DB_USER__"
```
Comment (autra) [inline conf/database.yml:5]: I won't block merging for that, but I take advantage of this PR to state again how I disagree with yunohost reinventing a template engine (which is what `ynh_replace_vars` is basically).  Yunohost should use *any* existing templating engine and call it a day (jinja, mustache, whatever...).

This `ynh_add_config` is always painful for me. I never remember how variables should be replaced and always need to look up the doc for its particular cases. Anyway, I may be speaking into the wind...

File: conf/diaspora_sidekiq.service [hunk 2]
```diff
@@ -3,29 +3,32 @@ Description=Diaspora social network (sidekiq - instance __APP__)
...
 ProtectSystem=full
 ProtectControlGroups=yes
-ProtectKernelModules=yes
-ProtectKernelTunables=yes
-LockPersonality=yes
-SystemCallFilter=~@clock @debug @module @mount @obsolete @reboot @setuid @swap
+#ProtectKernelModules=yes
+#ProtectKernelTunables=yes
+#LockPersonality=yes
+#SystemCallFilter=~@clock @debug @module @mount @obsolete @reboot @setuid @swap
```
Comment (autra) [inline conf/diaspora_sidekiq.service:31]: Can you elaborate why the changes in this file? You had trouble starting the server with the changes I did in 3e3cc64?

File: conf/diaspora_web.service [hunk 3]
```diff
@@ -18,7 +21,7 @@ Restart=always
 NoNewPrivileges=yes
 PrivateTmp=yes
 PrivateDevices=yes
-RestrictAddressFamilies=AF_UNIX AF_INET AF_INET6
+RestrictAddressFamilies=AF_UNIX AF_INET AF_INET6 AF_NETLINK
```
Comment (autra) [inline conf/diaspora_web.service:24]: By curiosity, do you know why we need AF_NETLINK here?

File: conf/mpapis@gmail.com.pgp [hunk 4]
```diff
@@ -1,1129 +0,0 @@
------BEGIN PGP PUBLIC KEY BLOCK-----
```
Comment (autra) [inline conf/mpapis@gmail.com.pgp:1]: thanks for the clean!

File: manifest.json [hunk 5]
```diff
@@ -6,7 +6,7 @@
         "en": "Distributed social networking service",
         "fr": "Service de réseau social distribué"
     },
-    "version": "0.7.17.0~ynh1",
+    "version": "0.7.18.1~ynh1",
```
Comment (autra) [inline manifest.json:9]: ok I'll test if the upgrade works correctly with real data.

File: scripts/_common.sh [hunk 6]
```diff
@@ -1,5 +1,47 @@
...
+
+# dependencies used by the app
+pkg_dependencies="cmake imagemagick ghostscript curl libmagickwand-dev git libpq-dev redis-server postgresql"
+build_pkg_dependencies="libffi-dev libgdbm-dev libncurses5-dev libyaml-dev pkg-config sqlite3 libgmp-dev libssl-dev"
+
+ruby_version=2.7
+nodejs_version=14
+bundler_version=2.1.4
+
+# Workaround for Mastodon on Bullseye
```
Comment (autra) [inline scripts/_common.sh]: I don't understand why mastodon is mentioned here.

File: scripts/backup [hunk 7]
```diff
@@ -23,42 +28,44 @@ ynh_print_info --message="Loading installation settings..."
...
-# BACKUP DIASPORA DATABASE
+# BACKUP THE APP MAIN DIR
 #=================================================
-ynh_print_info --message="Backup Diaspora DB..."
 
-db_pass=$(ynh_app_setting_get --app=$app --key=psqlpwd)
-dump_file="$final_path/backup/$app.dump"
-pg_dump -d "dbname=$app user=$app password=$db_pass host=localhost" -Fc -f $dump_file
-ynh_backup --src_path="$dump_file"
+ynh_backup --src_path="$final_path"
```
Comment (autra) [inline scripts/backup:43]: So before I was only backuping what diaspora advices: the db and the upload folder. Are we sure we want to backup the whole folder, including the code and dependencies? (maybe you've seen cases where it is useful?)

File: scripts/install [hunk 8]
```diff
@@ -1,189 +1,208 @@
...
+  ynh_exec_warn_less ynh_exec_as $app RAILS_ENV=production $ynh_ruby_load_path $ld_preload bin/rake assets:precompile
+  ynh_exec_warn_less ynh_exec_as $app RAILS_ENV=production $ynh_ruby_load_path $ld_preload bin/bundle exec rails console << END
+user = User.build({username: '$admin', email: '$admin_mail', password: '$password', password_confirmation: '$password' })
+user.save
+user.seed_aspects
+Role.add_admin user.person
+END
+popd
+
+ynh_exec_warn_less ynh_install_app_dependencies $pkg_dependencies
```
Comment (autra) [inline scripts/install]: Why are `$pkg_dependencies` installed again here? (they are already installed on line 62)

File: scripts/ynh_install_ruby__2 [hunk 9]
```diff
@@ -0,0 +1,310 @@
+#!/bin/bash
```
Comment (autra) [inline scripts/ynh_install_ruby__2:1]: This file comes from the experimental helpers, is that so? (so that I know what to check when I'll update it).

File: scripts/upgrade [hunk 10]
```diff
@@ -82,47 +90,155 @@ if gpg --list-keys piotr.kuczynski@gmail.com >/dev/null 2>&1; then
...
-  git fetch
-  git checkout $current_tag
-  popd
-  source ./bundle_app
+	ynh_script_progression --message="Upgrading source files..." --weight=1
+
+	tmpdir="$(mktemp -d)"
+	cp "$final_path/live/config/diaspora.yml" "$tmpdir/diaspora.yml"
+	cp "$final_path/live/config/database.yml" "$tmpdir/database.yml"
+	ynh_secure_remove --file="$final_path/live"
```
Comment (autra) [inline scripts/upgrade:125]: This directory contains the `public/uploads` directory. We cannot blindly delete it.

File: scripts/upgrade [hunk 11]
```diff
@@ -82,47 +90,155 @@ if gpg --list-keys piotr.kuczynski@gmail.com >/dev/null 2>&1; then
...
+
+#=================================================
+# SPECIFIC UPGRADE
 #=================================================
-# ADVERTISE SERVICE IN ADMIN PANEL
+# UPDATE A CONFIG FILE
 #=================================================
+ynh_script_progression --message="Updating a configuration file..." --weight=1
+
+ynh_add_config --template="../conf/diaspora.yml" --destination="$final_path/live/config/diaspora.yml"
```
Comment (autra) [inline scripts/upgrade:165]: I don't really understand why we kept the old config file above, just to replace it here. Is there something I missed?

File: scripts/install [hunk 12]
```diff
@@ -1,189 +1,208 @@
...
+chmod 400 "$final_path/live/config/diaspora.yml"
+chown $app:$app "$final_path/live/config/diaspora.yml"
+
+ynh_add_config --template="../conf/database.yml" --destination="$final_path/live/config/database.yml"
+chmod 400 "$final_path/live/config/database.yml"
+chown $app:$app "$final_path/live/config/database.yml"
 
 #=================================================
-# NGINX CONFIGURATION
+# BUILD APP
```
Comment (autra) [inline scripts/install:127]: In the same spirit as the templating engine remark, I with the standard way of packaging apps didn't involve so many code duplications... That's why I introduced `bundle_app` and `create_service` that this PR removes...

Comment (autra) [reply scripts/install:127]: To be more precise, I'm not asking to change anything here, it's just a remark.

**Review by autra — COMMENTED** [review]
Thanks @yalh76 for this work! I have a few questions or remark, and one main concern about the upload folder. 

Either you want to address these or I'll take over, let me know what you prefer :-)

File: scripts/_common.sh (hunk 6 above)
Comment (autra) [reply scripts/_common.sh]: Ok I've read the issue. It's for project using ruby 2.7 which is the case for us. I've added a comment explaining that. The workaround can probably be removed once we use ruby 3

File: scripts/install (hunk 8 above)
Comment (autra) [reply scripts/install]: Remove this occurrence and moved the autoremove near the first `ynh_install_app_dependencies`

File: scripts/ynh_install_ruby__2 (hunk 9 above)
Comment (autra) [reply scripts/ynh_install_ruby__2:1]: answer, yes.

File: conf/diaspora_web.service (hunk 3 above)
Comment (autra) [reply conf/diaspora_web.service:24]: Ok, I'm not sure we need it (because I'm not sure we really communicate with the kernel) but leaving it for now.

File: scripts/upgrade (hunk 11 above)
Comment (autra) [reply scripts/upgrade:165]: Keeping this, as it makes sense for upgrade : we deal with each upgrade case separately.

File: scripts/upgrade (hunk 10 above)
Comment (autra) [reply scripts/upgrade:125]: Just tested, and yup, the upgrade delete the uploads.

I think I'll just put the uploads folder outside of the "live" folder and link it there. That way it'll be untouched by the upgrade process, while still allowing us to trash the live folder whenever we want.

**Comment by autra** [comment]
!testme

**Comment by autra** [comment]
The test is failing but it's a timeout. I executed `package_check` locally and it passes, so let's merge that! Thanks @yalh76 for your work :-)
